man/fgjs.1.in
.\" Copyright (C) 2002 Cameron Moore
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
.\" Or try here: http://www.fsf.org/copyleft/gpl.html
.\"
.TH FGJS 1 "January 2002" "@VERSION@" "@PACKAGE@"
.SH NAME
fgjs \- FlightGear Joystick Utility
.SH SYNOPSIS
.B fgjs
.SH DESCRIPTION
.B fgjs
is a joystick utility for the FlightGear Flight Simulator.  It allows
you to generate an appropriate configuration file for your joystick.
For more information visit <http://wiki.flightgear.org/Joystick>.
.SH BUGS
Send bug reports to <flightgear\-devel@lists.sourceforge.net> or visit
<http://code.google.com/p/flightgear\-bugs>.
.SH SEE ALSO
fgfs(1), fgpanel(1), js_demo(1), terrasync(1)
.SH AUTHORS
Tony Peden, et al.  See
.I Thanks
file for a list of contributors.


NEWS
Version 2.0.0 - February 25, 2010

  FlightGear 2.0.0. reflects the maturation of the OpenSceneGraph port
  that started with the previous 1.9.0 release. In addition to many
  internal code improvements, FlightGear 2.0.0. marks the introduction
  of many new exciting improvements in the graphics and sound system,
  as well as improved usability of key features, and improved behavior
  of exsisting features. Highlights of this new version include:
  Dramatic new 3D clouds, dramatic lighting conditions, improved
  support for custom scenery, and many many new and detailed aircraft
  models.

  Sound 
  * Complete overhaul of the sound code
  * doppler effects
  * distance attenuation
  * 3D positional sound sources
  * assignment of sound sources to external objects (i.e. AI controlled
    aircraft)
  * User selection of the sound device

  Visual Effects
  * Use of Shaders for dynamic textures
  * Use of Effects files
  * Improved 3D clouds
  * Color changes based on humidity and other weather effects allow for very
    dramatic lighting conditions
  * Dynamic water textures
  * Text animation based on OSGText

  Usability
  * Allow screenshots in more common file formats
  * User selectable sound device
  * More intuitive selection of the weather settings through the GUI and/or
    commandline

  Infrastructure
  * Airport geometry data can be read from the scenery, allowing for more
    flexible regeneration of terrain tiles

  Internals
  * Improved efficiency of the property tree
  * A more efficient ground cache
  * Many improvements to the route management code
  * Removed many compiler warnings
  * More realistic atmosphere model

  Behavior
  * More realistic ILS behavior
  * Autopilot improvements
  * A generic autobrake function
  * Winds over mountainous areas cause up- and downdrafts that can be used
    for gliding
  * More realistic behavior of the route manager
  * Wild fires, which can be extinguished by firefighter aircraft operating
    across the multplayer server
  * Navaid frequencies and radials can be transmitted to Atlas

  Utilities
  * A python script to visualize Yasim configuration files in Blender

  AI
  * Allow traffic departing and arriving at the same airport
  * Add Ground Vehicles - including automobiles, trucks, articulated trucks,
    trains (including high speed trains)
  * ATC interactions between AI aircraft and ground controllers
  * Performance characteristics of AI aircraft can be specified in a
    performance database
  * Push-back vehicles are available for a selected number of aircraft
  * Add escorts for AI carrier - frigates, guided missile cruiser, amphibious
    warfare ships now make up the Vinson Battle Group
  * Improved radar functionality - now detects AI escorts etc.
  * AI objects are now solid (i.e. users can collide with them)
  * Some preliminary support for SID/STAR procedures for AI aircraft

Version 1.9.0
* December 20, 2008 (source code snapshot release)


New in 0.9.10
* April 5, 2006

* Many improvements to the multi-player mode.  Other airplanes are now
  updated smoothly.  There is a really nifty google maps based moving
  map that anyone can use to track all online activity.
* AI planes have improved ground following and are smarter about
  selecting the active runway.
* Include the ability to send ATC and AI aircraft communication to a
  local festival server to be converted to spoken speech output.
* Added a detailed model of the Honeywell MK VIII EGPWS ground
  proximity warning system.
* Complete overhaul/redo of terrain textures based on USGS satellite imagery.
* Added background ATC chatter.
* Updated/improved rain cone rendering.
* Updated manual.

* Upgraded to JSBSim v2.0
* Fix a discrepancy in sign of twist and incidence values in YASim
  between code and documentation.
* Updates to YASim for more fine grained propeller modeling.
* UFO will now honor elevator trim value so that it will respond to
  the autopilot altitude hold module.
* UFO can adjust it's speed range so you can use it for super high
  speed world exploration, or extremly fine grained precise positioning.
* Much work and progress with carrier interaction and launch/recovery
  modeling.

* Add a feature that allows the end user to click any where on the
  terrain.  The system will compute a ground intersection and report
  back the lon, lat, and elevation of the click point.  The click
  point is reported in the property system so it can be read by just
  about anything including external programs.  Later we added the
  ability to place an object at the click point as well.
* Add a feature to play one-off wav files.  These are qeued and played
  sequentially.  This was then leveraged to build an ATC background
  chatter module in nasal that can be turned on/off at run time.  This
  could also be leveraged for tutorials or other interactive type
  scripts to play prerecorded messages at specific times or under
  specific conditions..
* Allow clock to report local time as well as universal.
* Add a model of a KLN89 GPS.
* Allow nav radios to be slaved to the gps.
* Allow enabling/disabling of menu entries.
* Replace old hardcoded frame rate display with a much more flexible
  nasal/gui version.
* Add a listener function to properties so that nasal code can be
  triggered whenever a property changes.  This is more efficient than
  polling.
* Add a save-on-exit option to save user preferences.
* Add support for point sprites.
* Add support for emissive layers (such as lights) on 2d panels.
* Updated gui styles.
* Updated airport and navaid data.
* Added a new Nasal/XML based tutorial system.

* Fix a bug in the metar (real weather) threading that could lead to
  long pauses and even crashes.
* Various compiler/platform fixes.
* Fix various valgrind complaints.
* Better use of exceptions in places.
* Fix the 2d compass texture handling.
* Many tweaks and improvements to our keyboard input handling system.

src/Input/fgjs.cxx
// fgjs.cxx -- assign joystick axes to flightgear properties
//
// Updated to allow xml output & added a few bits & pieces
// Laurie Bradshaw, Jun 2005
//
// Written by Tony Peden, started May 2001
//
// Copyright (C) 2001  Tony Peden (apeden@earthlink.net)
// Copyright (C) 2006  Stefan Seifert
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <simgear/compiler.h>

#ifdef _WIN32
#  include <winsock2.h>
#endif

#include <math.h>

#include <iostream>
#include <fstream>
#include <string>

using std::fstream;
using std::cout;
using std::cin;
using std::endl;
using std::ios;
using std::string;

#include <simgear/constants.h>
#include <simgear/debug/logstream.hxx>
#include <simgear/misc/sg_path.hxx>
#include <simgear/misc/sgstream.hxx>
#include <simgear/structure/exception.hxx>
#include <simgear/props/props_io.hxx>

#include <Main/fg_io.hxx>
#include <Main/fg_props.hxx>
#include <Main/globals.hxx>

#include "jsinput.h"

#ifdef __APPLE__
#  include <CoreFoundation/CoreFoundation.h>
#endif

using simgear::PropertyList;

bool confirmAnswer() {
    char answer;
    do {
        cout << "Is this correct? (y/n) $ ";
        cin >> answer;
        cin.ignore(256, '\n');
        if (answer == 'y')
            return true;
        if (answer == 'n')
            return false;
    } while (true);
}

string getFGRoot( int argc, char *argv[] );

int main( int argc, char *argv[] ) {

    for (int i = 1; i < argc; i++) {
        if (strcmp("--help", argv[i]) == 0) {
            cout << "Usage:" << endl;
            cout << "  --help\t\t\tShow this help" << endl;
            exit(0);
	} else if (strncmp("--fg-root=", argv[i], 10) == 0) {
	    // used later
        } else {
            cout << "Unknown option \"" << argv[i] << "\"" << endl;
            exit(0);
        }
    }

    jsInit();

    jsSuper *jss = new jsSuper();
    jsInput *jsi = new jsInput(jss);
    jsi->displayValues(false);

    cout << "Found " << jss->getNumJoysticks() << " joystick(s)" << endl;

    if(jss->getNumJoysticks() <= 0) {
        cout << "Can't find any joysticks ..." << endl;
        exit(1);
    }
    cout << endl << "Now measuring the dead band of your joystick. The dead band is the area " << endl
                 << "where the joystick is centered and should not generate any input. Move all " << endl
                 << "axes around in this dead zone during the ten seconds this test will take." << endl;
    cout << "Press enter to continue." << endl;
    cin.ignore(1024, '\n');
    jsi->findDeadBand();
    cout << endl << "Dead band calibration finished. Press enter to start control assignment." << endl;
    cin.ignore(1024, '\n');

    jss->firstJoystick();
    fstream *xfs = new fstream[ jss->getNumJoysticks() ];
    SGPropertyNode_ptr *jstree = new SGPropertyNode_ptr[ jss->getNumJoysticks() ];
    do {
        cout << "Joystick #" << jss->getCurrentJoystickId()
             << " \"" << jss->getJoystick()->getName() << "\" has "
             << jss->getJoystick()->getNumAxes() << " axes" << endl;

        char filename[16];
        snprintf(filename, 16, "js%i.xml", jss->getCurrentJoystickId());
        xfs[ jss->getCurrentJoystickId() ].open(filename, ios::out);
        jstree[ jss->getCurrentJoystickId() ] = new SGPropertyNode();
    } while ( jss->nextJoystick() );

    SGPath templatefile( getFGRoot(argc, argv) );
    templatefile.append("Input");
    templatefile.append("Joysticks");
    templatefile.append("template.xml");

    SGPropertyNode *templatetree = new SGPropertyNode();
    try {
        readProperties(templatefile.str().c_str(), templatetree);
    } catch (sg_io_exception e) {
        cout << e.getFormattedMessage ();
    }

    PropertyList axes = templatetree->getChildren("axis");
    for(PropertyList::iterator iter = axes.begin(); iter != axes.end(); iter++) {
        cout << "Move the control you wish to use for " << (*iter)->getStringValue("desc")
             << " " << (*iter)->getStringValue("direction") << endl;
        cout << "Pressing a button skips this axis" << endl;
        fflush( stdout );
        jsi->getInput();
        if (jsi->getInputAxis() != -1) {
            cout << endl << "Assigned axis " << jsi->getInputAxis()
                 << " on joystick " << jsi->getInputJoystick()
                 << " to control " << (*iter)->getStringValue("desc") << endl;
            if ( confirmAnswer() ) {
                SGPropertyNode *axis = jstree[ jsi->getInputJoystick() ]->getChild("axis", jsi->getInputAxis(), true);
                copyProperties(*iter, axis);
                axis->setDoubleValue("dead-band", jss->getJoystick(jsi->getInputJoystick())
                        ->getDeadBand(jsi->getInputAxis()));
                axis->setDoubleValue("binding/factor", jsi->getInputAxisPositive() ? 1.0 : -1.0);
            } else {
                iter--;
            }
        } else {
            cout << "Skipping control" << endl;
            if ( ! confirmAnswer() )
                iter--;
        }
        cout << endl;
    }

    PropertyList buttons = templatetree->getChildren("button");
    for(PropertyList::iterator iter = buttons.begin(); iter != buttons.end(); iter++) {
        cout << "Press the button you wish to use for " << (*iter)->getStringValue("desc") << endl;
        cout << "Moving a joystick axis skips this button" << endl;
        fflush( stdout );
        jsi->getInput();
        if (jsi->getInputButton() != -1) {
            cout << endl << "Assigned button " << jsi->getInputButton()
                 << " on joystick " << jsi->getInputJoystick()
                 << " to control " << (*iter)->getStringValue("desc") << endl;
            if ( confirmAnswer() ) {
                SGPropertyNode *button = jstree[ jsi->getInputJoystick() ]->getChild("button", jsi->getInputButton(), true);
                copyProperties(*iter, button);
            } else {
                iter--;
            }
        } else {
            cout << "Skipping control" << endl;
            if (! confirmAnswer())
                iter--;
        }
        cout << endl;
    }

    cout << "Your joystick settings are in ";
    for (int i = 0; i < jss->getNumJoysticks(); i++) {
        try {
            cout << "js" << i << ".xml";
            if (i + 2 < jss->getNumJoysticks())
                cout << ", ";
            else if (i + 1 < jss->getNumJoysticks())
                cout << " and ";

            jstree[i]->setStringValue("name", jss->getJoystick(i)->getName());
            writeProperties(xfs[i], jstree[i], true);
        } catch (sg_io_exception e) {
            cout << e.getFormattedMessage ();
        }
        xfs[i].close();
    }
    cout << "." << endl << "Check and edit as desired. Once you are happy," << endl
         << "move relevant js<n>.xml files to $FG_ROOT/Input/Joysticks/ (if you didn't use" << endl
         << "an attached controller, you don't need to move the corresponding file)" << endl;

    delete jsi;
    delete[] xfs;
    delete jss;
    delete[] jstree;

    return 1;
}

char *homedir = ::getenv( "HOME" );
char *hostname = ::getenv( "HOSTNAME" );
bool free_hostname = false;

// Scan the command line options for the specified option and return
// the value.
static string fgScanForOption( const string& option, int argc, char **argv ) {
    int i = 1;

    if (hostname == NULL)
    {
        char _hostname[256];
        gethostname(_hostname, 256);
        hostname = strdup(_hostname);
        free_hostname = true;
    }

    SG_LOG(SG_INPUT, SG_INFO, "Scanning command line for: " << option );

    int len = option.length();

    while ( i < argc ) {
        SG_LOG( SG_INPUT, SG_DEBUG, "argv[" << i << "] = " << argv[i] );

        string arg = argv[i];
        if ( arg.find( option ) == 0 ) {
            return arg.substr( len );
        }

        i++;
    }

    return "";
}

// Scan the user config files for the specified option and return
// the value.
static string fgScanForOption( const string& option, const string& path ) {
    sg_gzifstream in( path );
    if ( !in.is_open() ) {
        return "";
    }

    SG_LOG( SG_INPUT, SG_INFO, "Scanning " << path << " for: " << option );

    int len = option.length();

    in >> skipcomment;
    while ( ! in.eof() ) {
        string line;
        getline( in, line, '\n' );

        // catch extraneous (DOS) line ending character
        if ( line[line.length() - 1] < 32 ) {
            line = line.substr( 0, line.length()-1 );
        }

        if ( line.find( option ) == 0 ) {
            return line.substr( len );
        }

        in >> skipcomment;
    }

    return "";
}

// Scan the user config files for the specified option and return
// the value.
static string fgScanForOption( const string& option ) {
    string arg("");

#if defined( unix ) || defined( __CYGWIN__ )
    // Next check home directory for .fgfsrc.hostname file
    if ( arg.empty() ) {
        if ( homedir != NULL ) {
            SGPath config( homedir );
            config.append( ".fgfsrc" );
            config.concat( "." );
            config.concat( hostname );
            arg = fgScanForOption( option, config.str() );
        }
    }
#endif

    // Next check home directory for .fgfsrc file
    if ( arg.empty() ) {
        if ( homedir != NULL ) {
            SGPath config( homedir );
            config.append( ".fgfsrc" );
            arg = fgScanForOption( option, config.str() );
        }
    }

    return arg;
}

// Read in configuration (files and command line options) but only set
// fg_root
string getFGRoot ( int argc, char **argv ) {
    string root;

    // First parse command line options looking for --fg-root=, this
    // will override anything specified in a config file
    root = fgScanForOption( "--fg-root=", argc, argv);

    // Check in one of the user configuration files.
    if (root.empty() )
        root = fgScanForOption( "--fg-root=" );

    // Next check if fg-root is set as an env variable
    if ( root.empty() ) {
        char *envp = ::getenv( "FG_ROOT" );
        if ( envp != NULL ) {
            root = envp;
        }
    }

    // Otherwise, default to a random compiled-in location if we can't
    // find fg-root any other way.
    if ( root.empty() ) {
#if defined( __CYGWIN__ )
        root = "../data";
#elif defined( _WIN32 )
        root = "..\\data";
#elif defined(__APPLE__) 
        /*
        The following code looks for the base package inside the application 
        bundle, in the standard Contents/Resources location. 
        */

        CFURLRef resourcesUrl = CFBundleCopyResourcesDirectoryURL(CFBundleGetMainBundle());

        // look for a 'data' subdir
        CFURLRef dataDir = CFURLCreateCopyAppendingPathComponent(NULL, resourcesUrl, CFSTR("data"), true);

        // now convert down to a path, and the a c-string
        CFStringRef path = CFURLCopyFileSystemPath(dataDir, kCFURLPOSIXPathStyle);
        root = CFStringGetCStringPtr(path, CFStringGetSystemEncoding());

        // tidy up.
        CFRelease(resourcesUrl);
        CFRelease(dataDir);
        CFRelease(path);
#else
        root = PKGLIBDIR;
#endif
    }

    SG_LOG(SG_INPUT, SG_INFO, "fg_root = " << root );

    return root;
}
